Add User Interface #154017417

wanjiku-carol opened the pull request:
**What does this PR do?**
Implements a user interface in React JS. Implements tests for React JS components

**Description of work done?**
- Divide the features into components
- Implement the component methods using states and React component methods
- Changes the tests to utilize the change to React implementation
- Add test utilities like enzyme, jsdom, and chai
- Add `npm start` to the README.

**How can this be manually tested?**
Install the relevant dependencies:
> $ npm install

Run the tests:
> $ npm test

**Relevant pivotal tracker stories?**
[#154017417](https://www.pivotaltracker.com/story/show/154017417)

**Screenshots:**
<img width="1255" alt="screen shot 2018-01-24 at 4 37 26 pm" src="https://user-images.githubusercontent.com/27045854/35334911-f9d9f338-0124-11e8-87ec-8d46ea020dc3.png">
<img width="1061" alt="screen shot 2018-01-24 at 4 37 39 pm" src="https://user-images.githubusercontent.com/27045854/35334915-fcfb65a6-0124-11e8-93eb-685bba12b6a1.png">

Commit:
ft(UI)Add a user interface

Commit:
ft(React)Implement React
- Create components and modify code to make use of React
- Adjust functions (formerly class methods) to work with components

Commit:
ft(return)Return result in the result text area
- Add new state
- Pass state to the result component

Commit:
ft(result)Print result in text area

Commit:
ft(tests)Add Component tests
- Set up jsDOM configurations
- Add enzyme, chai, jsdom, sidon and other test utilities
- Write component tests

Commit:
ft(tests)Add method tests
- Add tests for methods
- Ensure tests are passing

Commit:
ft(README)Add run app command
- Add the `npm start` command to the README file

Commit:
bg(travis)Remove react-addons-test-utils

Commit:
bg(travis)Change version in

Commit:
bg(travis)Extend travis timeout

Commit:
bg(travis)Set travis timeout

Commit:
bg(travis)Set travis timeout

Commit:
bg(travis)Set travis timeout

Commit:
bg(travis)Remove watch on tests

LWanjiru commented on app.js:
```diff
@@ -0,0 +1,23 @@
+
+}
+
+ReactDOM.render(<App />, document.getElementById("app"));
```
Add a new line here.

Also, instead of importing the whole of `ReactDOM`, consider importing just the `render` method from `react-dom`.

Commit:
ft(heroku)Add procfile and deploy to heroku

Commit:
bg(heroku)Run command on procfile

LWanjiru commented on app.js:
```diff
@@ -0,0 +1,23 @@
+import { Methods } from './src/components/Methods';
+import { Result } from './src/components/Result';
+
+class App extends React.Component{
```
Since this component doesn't have state, consider refactoring it to a pure function. 

This component can also be split into two smaller components for encapsulation. ie the `App` component that brings everything together, and a component, whose only purpose is to `render` whatever values have been passed into the `App` component.

LWanjiru commented on app.js:
```diff
@@ -0,0 +1,23 @@
+        super();
+    }
+    render(){
+            return(
```
It is good practice to always house the rendered items in a `<div>` tag.

LWanjiru commented on css/styles.css:
```diff
@@ -0,0 +1,64 @@
+    font-size: 15px;
+    height: 30px;
+    border-radius: 5px;
+}
```
Add new line here.

LWanjiru commented on server.js:
```diff
@@ -0,0 +1,5 @@
+var app = express();
+
+app.use(express.static(__dirname)); //serves the index.html
+app.listen(process.env.PORT || 3000); //listens on port 3000 -> http://localhost:3000/
```
Remember to end the file with an empty new line.

Look at the best practices for commenting on code in JS.

LWanjiru commented on spec/header.spec.js:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import Adapter from 'enzyme-adapter-react-16';
+import { configure, mount, shallow} from 'enzyme';
```
Import only methods that will be used in the test.

LWanjiru commented on spec/helpers/browser.js:
```diff
@@ -0,0 +1,9 @@
+require('babel-register')();
+
+var jsdom = require('jsdom');
```
I am curious about this file. What purpose does it serve?

wanjiku-carol replied:
This is the setup file for JSDOM. It is referred to in the test script in the `package.json`

LWanjiru commented on spec/methods.spec.js:
```diff
@@ -0,0 +1,48 @@
+        expect(wrapper.state().value).to.equal('hasVowel');
+    });
+
+    it('Button sets method for response', ()=>{
```
Love this test, it goes deeper. Should do the same with the other components, test beyond just what it can do.

Commit:
bg(heroku)Remove express from dev dependencies to dependencies

egichuri commented on spec/stringclass.spec.js:
```diff
@@ -1,162 +1,170 @@
-    const testVowels = new stringClass("We are happy").hasVowels();
-    it('returns true if vowel found', function(){
-        expect(testVowels).toBe(true);
+    it('Returns false if string has a vowel', function(){
```
i think this should be `doesn't have`

egichuri commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+function isQuestion(name){
+    /*Checks if string is a question */
+    if(wordCheck(name)=== true){
+       if(name.includes("?")){
```
Have you tested this against a phrase that has a `?` but isn't a question? e.g.  `tes?t`

egichuri commented on spec/stringclass.spec.js:
```diff
@@ -1,162 +1,170 @@
-        expect(arrayString.length).toEqual(3);
+    it('Returns the string as an array', function(){
+        const arrayString = stringMethods['isIn']("Yes");
+        expect(arrayString).to.be.an('array').that.includes('Y', 'e', 's' );
```
From this description - `Returns a list of the words in the string, as an Array.` - should the method return a list of letters or the word `Yes`?

egichuri commented on spec/stringclass.spec.js:
```diff
@@ -1,162 +1,170 @@
+describe('function alternateCase', function(){
+    it('Returns string in alternated case', function(){
+        const alternateString = stringMethods['alternatingCase']("Hello World");
+        expect(alternateString).to.equal("HeLlO WoRlD");
```
nitpick, but maybe this should take into account whitespaces and non-alpha characters.

egichuri commented on src/components/Methods.js:
```diff
@@ -0,0 +1,53 @@
+                        <p><b>Select method</b></p>
+                        <select
+                            className="select-options"
+                            onChange={event => this.setState({ value: event.target.value })}
```
you might need to set a default method for the initial state, or add `Select One` to the dropdown. i got an error with `hasVowels` if used before another option is selected since the value is set `onChange`. (hope that makes sense).

egichuri commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+        let vowels = ['a','e','i','o','u'];
+        let nameSplit = name.split("");
+        for(let word of nameSplit){
+            for(let vowel of vowels){
```
you could use `.includes` or `indexOf` instead of the nested loop + conditional. https://stackoverflow.com/a/1473742

wanjiku-carol replied:
The use of `.includes` or shortcut methods was forbidden for this method in the CP2 requirements

egichuri commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+                return name.split(" ");
+            } else{
+                return name.split("");
+            }
```
you can also `return name.includes(" ") ? name.split(" ") : name.split();`

i think `.split()` uses `""` by default

egichuri commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+            5:"five",6:"six",7:"seven",8:"eight",9:"nine"};
+            let splitNumber = name.toLocaleString().split(",");
+            let listNums = [];
+            for(let i=0; i<splitNumber.length; i++){
```
i'd change [line 197](https://github.com/wanjiku-carol/string-class/pull/6/files?diff=split#diff-db5888d39d0d4044481dcce4431e7c8cR197) to `let splitNumber = name.toString().split("");` and simplify this to 
```
 for(let i=0; i<splitNumber.length; i++){
       listNums.push(numbers[splitNumber[i]]);
}
```

egichuri reviewed (commented):
looks good! just a few comments for some recommendations.

LWanjiru commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+
+function wordCheck(name){
```
Look into ES6 arrow functions.

LWanjiru commented on src/components/stringclass.js:
```diff
@@ -0,0 +1,249 @@
+    
+}
+
+function ucFirst(name){
```
Consider using descriptive variable names ie `firstToUpper`. One of the advantages of doing this is that it makes your code self describing and reduces the amount of commenting done in the code.

wanjiku-carol replied:
`ucFirst` was the specific method name in the CP2 document.

LWanjiru commented on webpack.config.js:
```diff
@@ -13,13 +13,19 @@ var path = require('path');
                  loader: 'babel-loader',
                  query: {
-                     presets: ['es2015', 'react']
+                     presets: ['es2015','stage-0', 'react']
```
I noticed the presets exist both here and in the `.babelrc` file, do they serve different purposes?

LWanjiru reviewed (commented):
Great work so far. Look into the pointers I have shared and feel free to reach out for any clarifications. Keep going, you are on the right track. 
Kudos!

Commit:
ft(PR)Address requested changes in PR review
- Ensure the `isIn` method returns an array of words and not letters
- Ensure the `isQuestion` method returns true for the question mark at the end of the string/word and not in between a word
- Use arrow functions
- Set a default value method.
